torchvision/prototype/datasets/_builtin/lsun.py

```diff
@@ -0,0 +1,186 @@
+    Environment = Any
+
+
+class LmdbKeyExtractor(IterDataPipe[Tuple[str, bytes]]):
+    def __init__(self, datapipe: IterDataPipe[str]) -> None:
+        self.datapipe = datapipe
+
+    def __iter__(self) -> Iterator[Tuple[str, bytes]]:
+        import lmdb
+
+        for path in self.datapipe:
+            with lmdb.open(path, readonly=True) as env:
+                with env.begin(write=False) as txn:
+                    keys = b"\n".join(key for key in txn.cursor().iternext(keys=True, values=False))
+                    yield path, keys
+
+
+class LmdbLoader(IterDataPipe[Tuple[Environment, bytes]]):
+    def __init__(self, datapipe: IterDataPipe[str]) -> None:
+        self.datapipe = datapipe
+
+    def __iter__(self) -> Iterator[Tuple[Environment, bytes]]:  # type: ignore[valid-type]
+        import lmdb
+
+        for cache_path in self.datapipe:
+            env = lmdb.open(str(pathlib.Path(cache_path).parent), readonly=True)
+
+            with open(cache_path, "rb") as file:
+                for key in file:
+                    yield env, key.strip()
+
+
+class LmdbReader(IterDataPipe):
+    def __init__(self, datapipe: IterDataPipe[Tuple[Environment, bytes]]):
+        self.datapipe = datapipe
+
+    def __iter__(self) -> Iterator[Tuple[str, bytes, io.BytesIO]]:
+        for env, key in self.datapipe:
+            with env.begin(write=False) as txn:
+                yield env.path(), key, io.BytesIO(txn.get(key))
+
+
+class LsunHttpResource(HttpResource):
+    def __init__(self, *args: Any, extract: bool = True, **kwargs: Any) -> None:
+        super().__init__(*args, extract=extract, **kwargs)
+
+    def _loader(self, path: pathlib.Path) -> IterDataPipe[str]:
+        # LMDB datasets cannot be loaded through an open file handle, but have to be loaded through the path of the
+        # parent directory.
+        return IterableWrapper([str(next(path.rglob("data.mdb")).parent)])
+
+
+class Lsun(Dataset):
+    def _make_info(self) -> DatasetInfo:
+        return DatasetInfo(
+            "lsun",
+            type=DatasetType.IMAGE,
+            categories=(
+                "bedroom",
+                "bridge",
+                "church_outdoor",
+                "classroom",
+                "conference_room",
+                "dining_room",
+                "kitchen",
+                "living_room",
+                "restaurant",
+                "tower",
+            ),
+            valid_options=dict(split=("train", "val", "test")),
+            dependencies=("lmdb",),
+            homepage="https://www.yf.io/p/lsun",
+        )
+
+    _CHECKSUMS = {
+        ("train", "bedroom"): "",
+        ("train", "bridge"): "",
+        ("train", "church_outdoor"): "",
+        ("train", "classroom"): "",
+        ("train", "conference_room"): "",
+        ("train", "dining_room"): "",
+        ("train", "kitchen"): "",
+        ("train", "living_room"): "",
+        ("train", "restaurant"): "",
+        ("train", "tower"): "",
```

Comment:
The download is painfully slow. Would it be possible to spin up a cloud instance to download this and compute the checksums? After installing `torchvision` from this PR, it is enough to run

```py
import hashlib
import pathlib

from torchvision.prototype import datasets


next(iter(datasets.load("lsun", split="train", skip_integrity_check=True)))


def sha256sum(path, chunk_size=1024 * 1024):
    checksum = hashlib.sha256()
    with open(path, "rb") as f:
        for chunk in iter(lambda: f.read(chunk_size), b""):
            checksum.update(chunk)
    return checksum.hexdigest()


for path in pathlib.Path(datasets.home()).joinpath("lsun").glob("*_train_lmdb.zip"):
    print(path.name, sha256sum(path))
```

The data needs ~160 GB of storage.

Reply:
```suggestion
        ("train", "bedroom"): "a15644c2e7153106867f0adb3491fc41224102d1f01b24494fb47d9f5d1f174e",
        ("train", "bridge"): "2701b2d421bbc4d8a5e9b0652ff7c3b57cd6495da8e0e85c39533275b5a925a3",
        ("train", "church_outdoor"): "91128ae026840ac0c5982b4445ab5fc4e092d6847cca76793b2b1a0815c2e74a",
        ("train", "classroom"): "73a8a3e318819e1cc602f229673c3f51a68f3ece61e3764ce22df6abea4d0873",
        ("train", "conference_room"): "fa0a4cf72e7acfb103392eaf33640d5508a728e971f31877b01d64e1bde6068c",
        ("train", "dining_room"): "e4ee24c7c309360c3bf019123ce5bbf17434b2ba33abec2b0b07cfae715a52cb",
        ("train", "kitchen"): "b1993cf639aece5d207a27eb9ff872bcca9dff6472d8227a052c79d40ee753c4",
        ("train", "living_room"): "bd2c52b812c80f73ce3062a221396d13a52b5cce2f813b4cdf61937651281d7a",
        ("train", "restaurant"): "11a0a924b960cd0900e9b7477d684fb338bd99cc5f72db1caac592e4f497e09a",
        ("train", "tower"): "440caec74c9641cb51fd235a5970c8e4931c3af875aeb044a8bcca956e106309",
```